{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Coisas de setup pra poder rodar as funçoes, declaracoes de variaveis e etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_col<-\"Gene.symbol\"\n",
    "logFC_col<-\"logFC\"\n",
    "pvalue_col<-\"P.Value\"\n",
    "min_genes<-100\n",
    "max_genes<-1500\n",
    "file_in <- \"GSE49757_Septic_vs_Healthy.txt\"\n",
    "gmt_file <- \"Reactome_2016_15and100Genes.gmt\"\n",
    "deg_list <- read.csv(file_in, header = TRUE, sep = \"\\t\")\n",
    "deg_list <- deg_list[which(deg_list[, gene_col]!=\"\"), ]\n",
    "    ## Make logFC and p-value columns numeric\n",
    "deg_list[, logFC_col] <- as.numeric(deg_list[, logFC_col])\n",
    "deg_list[, pvalue_col] <- as.numeric(deg_list[, pvalue_col])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Declarando funções"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    ".get_pathway <- function(merge_p, term2gene, all_genes, deg_list,\n",
    "                        gene_col, logFC_col, pvalue_col, direction,\n",
    "                        min_genes, max_genes, p_cut){\n",
    "\n",
    "   \n",
    "    \n",
    "    if(tolower(direction) == \"up\"){\n",
    "        top <- deg_list[head(order(deg_list[, logFC_col], decreasing=TRUE), n=max_genes), ]\n",
    "    }else if(tolower(direction) == \"down\"){\n",
    "        top <- deg_list[head(order(deg_list[, logFC_col], decreasing=FALSE), n=max_genes), ]\n",
    "    }else if(tolower(direction) == \"any\"){\n",
    "        top <- deg_list[head(order(abs(deg_list[, logFC_col]), decreasing=FALSE), n=max_genes), ]\n",
    "    }else{\n",
    "        stop(\"Invalid direction argument\")\n",
    "    }\n",
    "\n",
    "    # add pi_value\n",
    "    top$pi_value = abs(top[, logFC_col])*-log10(top[, pvalue_col])\n",
    "\n",
    "    # order pi_value\n",
    "    top <- top[order(top$pi_value, decreasing=TRUE), ]\n",
    "    \n",
    "  \n",
    "    \n",
    "    for (i in seq(from=min_genes, to=max_genes, by=50)) {\n",
    "        top_genes  <- as.character(top[1:i, gene_col])\n",
    "        pathG <- .run_enrich(top_genes, all_genes, term2gene)\n",
    "        \n",
    "        ############################################\n",
    "        #    Parei aqui\n",
    "        ############################################\n",
    "        \n",
    "    \n",
    "        colnames(pathG) <- c(\"term\",  i)\n",
    "        return(pathG)\n",
    "        merge_p <- merge(merge_p, pathG, by=\"term\", all=TRUE)\n",
    "        return (merge_p)\n",
    "        merge_p[is.na(merge_p)] <- 1\n",
    "    }\n",
    "    rownames(merge_p) <- merge_p[, 1]\n",
    "    merge_p           <- merge_p[, -1]\n",
    "    merge_p2          <- log10(merge_p)*-1\n",
    "\n",
    "    path_cut_p <- log10(p_cut)*-1\n",
    "\n",
    "    df <- data.frame(matrix(0, nrow(merge_p2), ncol=0))\n",
    "    rownames(df) <- rownames(merge_p2)\n",
    "    #top cut with maximum MinuslogP\n",
    "    df$NG <- as.numeric(colnames(merge_p2)[apply(merge_p2, 1, which.max)])\n",
    "    df$p  <- as.numeric(apply(merge_p2, 1, max))\n",
    "    df$P  <- as.numeric(apply(merge_p2, 1, sum))\n",
    "\n",
    "    #How many columns above path_cut_p (freq)\n",
    "    df$times <- as.numeric(apply(merge_p2, 1, function(x) length(which(x > path_cut_p))))/ncol(merge_p2)\n",
    "\n",
    "    #If the pathway has times > 0\n",
    "    #First column above path_cut_p\n",
    "    df$first <- apply(merge_p2, 1, function(x) ifelse (length(which(x > path_cut_p)) >0,\n",
    "                                                as.numeric(colnames(merge_p2)[min(which(x > path_cut_p))]),\n",
    "                                                0))\n",
    "\n",
    "    #ES3\n",
    "    df$ES3 <- (1 - exp(-df$p))/(1 + 0.1*sqrt(df$NG))\n",
    "\n",
    "    #order\n",
    "    merge_p2 <- merge_p2[order(df[, \"first\"], decreasing=TRUE), ]\n",
    "\n",
    "    colnames(df) <- c(paste(\"TopCut_highestMinuslogP\", \"_\", direction, sep=\"\"),\n",
    "                      paste(\"maximum_MinuslogP\", \"_\", direction, sep=\"\"),\n",
    "                      paste(\"sum_MinuslogP\", \"_\", direction, sep=\"\"),\n",
    "                      paste(\"times_significant\", \"_\", direction, sep=\"\"),\n",
    "                      paste(\"FirstTopCut_significant\", \"_\", direction, sep=\"\"),\n",
    "                      paste(\"PEBBA_score\", \"_\", direction, sep=\"\"))\n",
    "\n",
    "    newList <- list(\"data.frame\" = df, \"data.frame\" = merge_p2)\n",
    "    return(newList)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "hide_input": true
   },
   "outputs": [],
   "source": [
    ".cutoff_path <- function(path_table, p_cut, direction){\n",
    "    df <- data.frame(matrix(0, nrow=ncol(path_table), ncol=0))\n",
    "    rownames(df) <- colnames(path_table)\n",
    "\n",
    "    df$MaxR  <- as.numeric(apply(path_table, 2, max))\n",
    "    df$SumR  <- as.numeric(apply(path_table, 2, sum))\n",
    "    path_cut_p <- log10(p_cut)*-1\n",
    "    #How many pathways above path_cut_p (freq)\n",
    "    df$times <- as.numeric(apply(path_table, 2,\n",
    "                               function(x) length(which(x > path_cut_p))))/nrow(path_table)\n",
    "    colnames(df) <- c(paste0(\"maximum_MinuslogP_\", direction),\n",
    "                    paste0(\"sum_MinuslogP_\", direction),\n",
    "                    paste0(\"times_significant_\", direction))\n",
    "    return(df)\n",
    "}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "hide_input": true
   },
   "outputs": [],
   "source": [
    ".run_enrich <- function(top_genes, all_genes, term2gene){\n",
    "    enriched <- as.data.frame(clusterProfiler::enricher(gene = top_genes,\n",
    "                                    pvalueCutoff = 1,\n",
    "                                    minGSSize = 1,\n",
    "                                    universe = all_genes,\n",
    "                                    TERM2GENE = term2gene,\n",
    "                                    qvalueCutoff = 1,\n",
    "                                    maxGSSize = 100))[, c(1, 6)]\n",
    "    return(enriched)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main e etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "hide_input": false
   },
   "outputs": [],
   "source": [
    "\n",
    "pebba <- function(file_in, gmt_file, gene_col=\"Gene.symbol\",\n",
    "                  logFC_col=\"logFC\", pvalue_col=\"P.Value\",\n",
    "                  min_genes=100, max_genes=1500,\n",
    "                  p_cut=0.2, verbose=TRUE,\n",
    "                  analysis_name=NULL, results_dir=\"Results\",\n",
    "                  force=FALSE){\n",
    "\n",
    "    # Validating inputs\n",
    "    if(min_genes < 50 | min_genes > 2900){\n",
    "        stop(\"Variable min_genes must be between 50 and 2900 genes\")\n",
    "    }\n",
    "    if(max_genes < 100 | max_genes > 3000){\n",
    "        stop(\"Variable max_genes must be between 100 and 3000 genes\")\n",
    "    }\n",
    "    if(p_cut < 0.00001 | p_cut > 1){\n",
    "        stop(\"Variable p_cut must be between 0.00001 and 1\")\n",
    "    }\n",
    "\n",
    "\n",
    "    # Preparing files and workspace\n",
    "    ## Disable scientifc notation\n",
    "    options(scipen=999)\n",
    "\n",
    "    ## Create a results directory\n",
    "    if(dir.exists(results_dir)){\n",
    "        if(!force){\n",
    "            stop(\"Stopping analysis: \", results_dir,\n",
    "                 \" already exists! Use force=TRUE to overwrite.\")\n",
    "        }\n",
    "    }else{\n",
    "        dir.create(results_dir)\n",
    "        dir.create(file.path(results_dir, \"Tables\"))\n",
    "        dir.create(file.path(results_dir, \"Heatmaps\"))\n",
    "    }\n",
    "\n",
    "    if(is.null(analysis_name)){\n",
    "        analysis_name <- \"PEBBA_analysis\"\n",
    "    }\n",
    "\n",
    "    ## Get information from all unique terms\n",
    "    gmt_res <- read_gmt_hier(gmt_file)\n",
    "    term2gene <- gmt_res[[1]]\n",
    "    path_desc <- gmt_res[[2]]\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    merge_p <- data.frame(unique(term2gene[1]))\n",
    "    #return( merge_p)\n",
    "    \n",
    "    if(is.character(file_in)){\n",
    "        deg_list <- read.csv(file_in, header = TRUE, sep = \"\\t\")\n",
    "        if(is.null(analysis_name)){\n",
    "            analysis_name <- tools::file_path_sans_ext(basename(file_in))\n",
    "        }\n",
    "    }else if(is.data.frame(file_in)){\n",
    "        deg_list <- file_in\n",
    "    }\n",
    "        \n",
    "         \n",
    "  \n",
    "\n",
    "    ## Remove rows that do not have a valid gene symbol\n",
    "    deg_list <- deg_list[which(deg_list[, gene_col]!=\"\"), ]\n",
    "    ## Make logFC and p-value columns numeric\n",
    "    deg_list[, logFC_col] <- as.numeric(deg_list[, logFC_col])\n",
    "    deg_list[, pvalue_col] <- as.numeric(deg_list[, pvalue_col])\n",
    "\n",
    "    ## Get background genes as a character vector\n",
    "    ## Empty values (non-annotated genes) will be removed\n",
    "    all_genes <- as.character(deg_list[, gene_col])\n",
    "\n",
    "    # Get cutoff values -------------------------------------------------------\n",
    "\n",
    "    if(verbose) message(\"Getting cutoff\")\n",
    "    ## Get info about p-value and log2fc cutoff used on each top segments\n",
    "    \n",
    "   \n",
    "    \n",
    "    table_cut <- .get_cutoff(deg_list, logFC_col, pvalue_col, min_genes, max_genes)\n",
    "\n",
    "     # ---------------------------PAREI AQUI------------------------------------------------ #\n",
    "    \n",
    "    \n",
    "    dirs <- c(\"up\", \"down\", \"any\")\n",
    "\n",
    "    cut_path_list <- lapply(dirs, function(direction){\n",
    "        if(verbose) message(direction)\n",
    "        if(verbose) message(\"Getting pathways\")\n",
    "        list_p <- .get_pathway(merge_p, term2gene, all_genes,\n",
    "                            deg_list, gene_col, logFC_col,\n",
    "                            pvalue_col, direction,\n",
    "                            min_genes, max_genes, p_cut)\n",
    "        return(list_p)\n",
    "\n",
    "        df <- list_p[[1]]\n",
    "        path <- list_p[[2]]\n",
    "\n",
    "        if(verbose) message(\"Getting pathway cutoff\")\n",
    "        cut_path <- .cutoff_path(path, p_cut, direction)\n",
    "        res <- list(cut_path, df, path)\n",
    "        names(res) <- c(\"cut_path\", \"df\", \"path\")\n",
    "        res\n",
    "    })\n",
    "    names(cut_path_list) <- dirs\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    ".get_cutoff <- function(deg_list, logFC_col, pvalue_col, min_genes, max_genes){\n",
    "    dirs <- c(\"down\", \"up\")\n",
    "\n",
    "    res <- lapply(dirs, function(direction){\n",
    "\n",
    "        decreasing <- ifelse(direction == \"down\", FALSE, TRUE)\n",
    "\n",
    "        top <- deg_list[head(order(deg_list[, logFC_col],\n",
    "                                   decreasing=decreasing),\n",
    "                             n=max_genes),\n",
    "                        c(logFC_col, pvalue_col)]\n",
    "        #Add pi_value\n",
    "        top$pi_value <- abs(top[, logFC_col]) * -log10(top[, pvalue_col])\n",
    "        #Order pi_value\n",
    "        top <- top[order(top$pi_value, decreasing=TRUE), ]\n",
    "        df1 <- data.frame(minFC=numeric(0), minP=numeric(0), minPi=numeric(0))\n",
    "        for (i in seq(from=min_genes, to=max_genes, by=50)) {\n",
    "            top_genes  <- top[1:i, ]\n",
    "            minFC <- min(abs(top_genes[, 1]))\n",
    "            maxP  <- max(top_genes[, 2])\n",
    "            minP  <- -log10(maxP)\n",
    "            minPi <- min(top_genes[i, 3])\n",
    "            rowX  <- data.frame(minFC=minFC, minP=minP, minPi=minPi)\n",
    "            df1 <- rbind(df1,rowX)\n",
    "        }\n",
    "        df1\n",
    "    })\n",
    "    names(res) <- dirs\n",
    "    top_cut <- seq(from=min_genes, to=max_genes, by=50)\n",
    "    res <- do.call(\"cbind\", res)\n",
    "    res <- cbind(top_cut, res)\n",
    "\n",
    "    res$fc <- apply(res, 1, function(x) min(x[2], x[6]) )\n",
    "    res$p  <- apply(res, 1, function(x) min(x[3], x[7]) )\n",
    "    res$pi <- apply(res, 1, function(x) min(x[4], x[8]) )\n",
    "\n",
    "    names(res) <- c(\"TopCut\", \"minimum_log2fc_down\", \"minimum_MinuslogP_down\",\n",
    "                    \"minimum_Pi_down\", \"minimum_log2fc_up\", \"minimum_MinuslogP_up\",\n",
    "                    \"minimum_Pi_up\", \"minimum_log2fc_combined\",\n",
    "                    \"minimum_MinuslogP_combined\", \"minimum_Pi_combined\")\n",
    "\n",
    "    rownames(res) <- res[, 1]\n",
    "    return(res)\n",
    "}\n",
    "\n",
    "read_gmt_hier <- function(fname){\n",
    "    res <- list(genes=list(), desc=list())\n",
    "    gmt <- file(fname)\n",
    "    gmt_lines <- readLines(gmt)\n",
    "    close(gmt)\n",
    "    gmt_list <- lapply(gmt_lines, function(x) unlist(strsplit(x, split=\"\\t\")))\n",
    "    gmt_names <- sapply(gmt_list, '[', 1)\n",
    "    gmt_desc <- lapply(gmt_list, '[', 2)\n",
    "    gmt_genes <- lapply(gmt_list, function(x){x[3:length(x)]})\n",
    "    names(gmt_desc) <- names(gmt_genes) <- gmt_names\n",
    "    res <- do.call(rbind, lapply(names(gmt_genes),\n",
    "                                 function(n) cbind.data.frame(term=n, gene=gmt_genes[[n]], hier=gmt_desc[[n]], stringsAsFactors=FALSE)))\n",
    "    res$term <- as.factor(res$term)\n",
    "    path_desc <- as.data.frame(cbind(gmt_names, gmt_desc))\n",
    "    return(list(res, path_desc))\n",
    "}                                  \n",
    "                                   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Testes e chamando uma função de cada vez"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# merge_P"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>term</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>Fertilization                                                              </td></tr>\n",
       "\t<tr><th scope=row>27</th><td>Reproduction                                                               </td></tr>\n",
       "\t<tr><th scope=row>53</th><td>Miscellaneous transport and binding events                                 </td></tr>\n",
       "\t<tr><th scope=row>71</th><td>Uptake and actions of bacterial toxins                                     </td></tr>\n",
       "\t<tr><th scope=row>98</th><td>Dopamine Neurotransmitter Release Cycle                                    </td></tr>\n",
       "\t<tr><th scope=row>121</th><td>Serotonin Neurotransmitter Release Cycle                                   </td></tr>\n",
       "\t<tr><th scope=row>139</th><td>Norepinephrine Neurotransmitter Release Cycle                              </td></tr>\n",
       "\t<tr><th scope=row>157</th><td>Acetylcholine Neurotransmitter Release Cycle                               </td></tr>\n",
       "\t<tr><th scope=row>174</th><td>Glutamate Neurotransmitter Release Cycle                                   </td></tr>\n",
       "\t<tr><th scope=row>198</th><td>Neurotransmitter Release Cycle                                             </td></tr>\n",
       "\t<tr><th scope=row>249</th><td>GABA synthesis release reuptake and degradation                            </td></tr>\n",
       "\t<tr><th scope=row>268</th><td>Na+slashCl_ dependent neurotransmitter transporters                        </td></tr>\n",
       "\t<tr><th scope=row>287</th><td>Amine compound SLC transporters                                            </td></tr>\n",
       "\t<tr><th scope=row>317</th><td>Transport of inorganic cationsslashanions and amino acidsslasholigopeptides</td></tr>\n",
       "\t<tr><th scope=row>416</th><td>Amino acid transport across the plasma membrane                            </td></tr>\n",
       "\t<tr><th scope=row>447</th><td>Amino acid and oligopeptide SLC transporters                               </td></tr>\n",
       "\t<tr><th scope=row>496</th><td>Zinc transporters                                                          </td></tr>\n",
       "\t<tr><th scope=row>513</th><td>Metal ion SLC transporters                                                 </td></tr>\n",
       "\t<tr><th scope=row>538</th><td>Amino acid synthesis and interconversion transamination                    </td></tr>\n",
       "\t<tr><th scope=row>563</th><td>Transport of vitamins nucleosides and related molecules                    </td></tr>\n",
       "\t<tr><th scope=row>601</th><td>Disorders of transmembrane transporters                                    </td></tr>\n",
       "\t<tr><th scope=row>616</th><td>Surfactant metabolism                                                      </td></tr>\n",
       "\t<tr><th scope=row>645</th><td>ABC transporters in lipid homeostasis                                      </td></tr>\n",
       "\t<tr><th scope=row>662</th><td>ABC_family proteins mediated transport                                     </td></tr>\n",
       "\t<tr><th scope=row>707</th><td>Signaling by Retinoic Acid                                                 </td></tr>\n",
       "\t<tr><th scope=row>749</th><td>RA biosynthesis pathway                                                    </td></tr>\n",
       "\t<tr><th scope=row>771</th><td>The canonical retinoid cycle in rods twilight vision                       </td></tr>\n",
       "\t<tr><th scope=row>794</th><td>Visual phototransduction                                                   </td></tr>\n",
       "\t<tr><th scope=row>891</th><td>Retinoid metabolism and transport                                          </td></tr>\n",
       "\t<tr><th scope=row>933</th><td>Metabolism of fat_soluble vitamins                                         </td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>32704</th><td>Phase 1 _ Functionalization of compounds                                 </td></tr>\n",
       "\t<tr><th scope=row>32796</th><td>Xenobiotics                                                              </td></tr>\n",
       "\t<tr><th scope=row>32816</th><td>Aflatoxin activation and detoxification                                  </td></tr>\n",
       "\t<tr><th scope=row>32837</th><td>Glutathione conjugation                                                  </td></tr>\n",
       "\t<tr><th scope=row>32875</th><td>Glucuronidation                                                          </td></tr>\n",
       "\t<tr><th scope=row>32895</th><td>Cytosolic sulfonation of small molecules                                 </td></tr>\n",
       "\t<tr><th scope=row>32913</th><td>Synthesis and interconversion of nucleotide di_ and triphosphates        </td></tr>\n",
       "\t<tr><th scope=row>32938</th><td>Detoxification of Reactive Oxygen Species                                </td></tr>\n",
       "\t<tr><th scope=row>32971</th><td>Metabolism of porphyrins                                                 </td></tr>\n",
       "\t<tr><th scope=row>32988</th><td>Purine metabolism                                                        </td></tr>\n",
       "\t<tr><th scope=row>33022</th><td>Metabolism of nucleotides                                                </td></tr>\n",
       "\t<tr><th scope=row>33109</th><td>Pyrimidine metabolism                                                    </td></tr>\n",
       "\t<tr><th scope=row>33133</th><td>Complex I biogenesis                                                     </td></tr>\n",
       "\t<tr><th scope=row>33182</th><td>Respiratory electron transport                                           </td></tr>\n",
       "\t<tr><th scope=row>33270</th><td>Formation of ATP by chemiosmotic coupling                                </td></tr>\n",
       "\t<tr><th scope=row>33286</th><td>Glyoxylate metabolism and glycine degradation                            </td></tr>\n",
       "\t<tr><th scope=row>33311</th><td>Pyruvate metabolism                                                      </td></tr>\n",
       "\t<tr><th scope=row>33338</th><td>Pyruvate metabolism and Citric Acid TCA cycle                            </td></tr>\n",
       "\t<tr><th scope=row>33386</th><td>Citric acid cycle TCA cycle                                              </td></tr>\n",
       "\t<tr><th scope=row>33405</th><td>Histidine lysine phenylalanine tyrosine proline and tryptophan catabolism</td></tr>\n",
       "\t<tr><th scope=row>33446</th><td>Mitochondrial tRNA aminoacylation                                        </td></tr>\n",
       "\t<tr><th scope=row>33467</th><td>tRNA Aminoacylation                                                      </td></tr>\n",
       "\t<tr><th scope=row>33509</th><td>Cytosolic tRNA aminoacylation                                            </td></tr>\n",
       "\t<tr><th scope=row>33533</th><td>tRNA modification in the nucleus and cytosol                             </td></tr>\n",
       "\t<tr><th scope=row>33572</th><td>Branched_chain amino acid catabolism                                     </td></tr>\n",
       "\t<tr><th scope=row>33595</th><td>Phase 4 _ resting membrane potential                                     </td></tr>\n",
       "\t<tr><th scope=row>33614</th><td>Mitochondrial translation initiation                                     </td></tr>\n",
       "\t<tr><th scope=row>33698</th><td>Mitochondrial translation termination                                    </td></tr>\n",
       "\t<tr><th scope=row>33782</th><td>Mitochondrial translation                                                </td></tr>\n",
       "\t<tr><th scope=row>33872</th><td>Mitochondrial translation elongation                                     </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|l}\n",
       "  & term\\\\\n",
       "\\hline\n",
       "\t1 & Fertilization                                                              \\\\\n",
       "\t27 & Reproduction                                                               \\\\\n",
       "\t53 & Miscellaneous transport and binding events                                 \\\\\n",
       "\t71 & Uptake and actions of bacterial toxins                                     \\\\\n",
       "\t98 & Dopamine Neurotransmitter Release Cycle                                    \\\\\n",
       "\t121 & Serotonin Neurotransmitter Release Cycle                                   \\\\\n",
       "\t139 & Norepinephrine Neurotransmitter Release Cycle                              \\\\\n",
       "\t157 & Acetylcholine Neurotransmitter Release Cycle                               \\\\\n",
       "\t174 & Glutamate Neurotransmitter Release Cycle                                   \\\\\n",
       "\t198 & Neurotransmitter Release Cycle                                             \\\\\n",
       "\t249 & GABA synthesis release reuptake and degradation                            \\\\\n",
       "\t268 & Na+slashCl\\_ dependent neurotransmitter transporters                        \\\\\n",
       "\t287 & Amine compound SLC transporters                                            \\\\\n",
       "\t317 & Transport of inorganic cationsslashanions and amino acidsslasholigopeptides\\\\\n",
       "\t416 & Amino acid transport across the plasma membrane                            \\\\\n",
       "\t447 & Amino acid and oligopeptide SLC transporters                               \\\\\n",
       "\t496 & Zinc transporters                                                          \\\\\n",
       "\t513 & Metal ion SLC transporters                                                 \\\\\n",
       "\t538 & Amino acid synthesis and interconversion transamination                    \\\\\n",
       "\t563 & Transport of vitamins nucleosides and related molecules                    \\\\\n",
       "\t601 & Disorders of transmembrane transporters                                    \\\\\n",
       "\t616 & Surfactant metabolism                                                      \\\\\n",
       "\t645 & ABC transporters in lipid homeostasis                                      \\\\\n",
       "\t662 & ABC\\_family proteins mediated transport                                     \\\\\n",
       "\t707 & Signaling by Retinoic Acid                                                 \\\\\n",
       "\t749 & RA biosynthesis pathway                                                    \\\\\n",
       "\t771 & The canonical retinoid cycle in rods twilight vision                       \\\\\n",
       "\t794 & Visual phototransduction                                                   \\\\\n",
       "\t891 & Retinoid metabolism and transport                                          \\\\\n",
       "\t933 & Metabolism of fat\\_soluble vitamins                                         \\\\\n",
       "\t⋮ & ⋮\\\\\n",
       "\t32704 & Phase 1 \\_ Functionalization of compounds                                 \\\\\n",
       "\t32796 & Xenobiotics                                                              \\\\\n",
       "\t32816 & Aflatoxin activation and detoxification                                  \\\\\n",
       "\t32837 & Glutathione conjugation                                                  \\\\\n",
       "\t32875 & Glucuronidation                                                          \\\\\n",
       "\t32895 & Cytosolic sulfonation of small molecules                                 \\\\\n",
       "\t32913 & Synthesis and interconversion of nucleotide di\\_ and triphosphates        \\\\\n",
       "\t32938 & Detoxification of Reactive Oxygen Species                                \\\\\n",
       "\t32971 & Metabolism of porphyrins                                                 \\\\\n",
       "\t32988 & Purine metabolism                                                        \\\\\n",
       "\t33022 & Metabolism of nucleotides                                                \\\\\n",
       "\t33109 & Pyrimidine metabolism                                                    \\\\\n",
       "\t33133 & Complex I biogenesis                                                     \\\\\n",
       "\t33182 & Respiratory electron transport                                           \\\\\n",
       "\t33270 & Formation of ATP by chemiosmotic coupling                                \\\\\n",
       "\t33286 & Glyoxylate metabolism and glycine degradation                            \\\\\n",
       "\t33311 & Pyruvate metabolism                                                      \\\\\n",
       "\t33338 & Pyruvate metabolism and Citric Acid TCA cycle                            \\\\\n",
       "\t33386 & Citric acid cycle TCA cycle                                              \\\\\n",
       "\t33405 & Histidine lysine phenylalanine tyrosine proline and tryptophan catabolism\\\\\n",
       "\t33446 & Mitochondrial tRNA aminoacylation                                        \\\\\n",
       "\t33467 & tRNA Aminoacylation                                                      \\\\\n",
       "\t33509 & Cytosolic tRNA aminoacylation                                            \\\\\n",
       "\t33533 & tRNA modification in the nucleus and cytosol                             \\\\\n",
       "\t33572 & Branched\\_chain amino acid catabolism                                     \\\\\n",
       "\t33595 & Phase 4 \\_ resting membrane potential                                     \\\\\n",
       "\t33614 & Mitochondrial translation initiation                                     \\\\\n",
       "\t33698 & Mitochondrial translation termination                                    \\\\\n",
       "\t33782 & Mitochondrial translation                                                \\\\\n",
       "\t33872 & Mitochondrial translation elongation                                     \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | term | \n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | Fertilization                                                               | \n",
       "| 27 | Reproduction                                                                | \n",
       "| 53 | Miscellaneous transport and binding events                                  | \n",
       "| 71 | Uptake and actions of bacterial toxins                                      | \n",
       "| 98 | Dopamine Neurotransmitter Release Cycle                                     | \n",
       "| 121 | Serotonin Neurotransmitter Release Cycle                                    | \n",
       "| 139 | Norepinephrine Neurotransmitter Release Cycle                               | \n",
       "| 157 | Acetylcholine Neurotransmitter Release Cycle                                | \n",
       "| 174 | Glutamate Neurotransmitter Release Cycle                                    | \n",
       "| 198 | Neurotransmitter Release Cycle                                              | \n",
       "| 249 | GABA synthesis release reuptake and degradation                             | \n",
       "| 268 | Na+slashCl_ dependent neurotransmitter transporters                         | \n",
       "| 287 | Amine compound SLC transporters                                             | \n",
       "| 317 | Transport of inorganic cationsslashanions and amino acidsslasholigopeptides | \n",
       "| 416 | Amino acid transport across the plasma membrane                             | \n",
       "| 447 | Amino acid and oligopeptide SLC transporters                                | \n",
       "| 496 | Zinc transporters                                                           | \n",
       "| 513 | Metal ion SLC transporters                                                  | \n",
       "| 538 | Amino acid synthesis and interconversion transamination                     | \n",
       "| 563 | Transport of vitamins nucleosides and related molecules                     | \n",
       "| 601 | Disorders of transmembrane transporters                                     | \n",
       "| 616 | Surfactant metabolism                                                       | \n",
       "| 645 | ABC transporters in lipid homeostasis                                       | \n",
       "| 662 | ABC_family proteins mediated transport                                      | \n",
       "| 707 | Signaling by Retinoic Acid                                                  | \n",
       "| 749 | RA biosynthesis pathway                                                     | \n",
       "| 771 | The canonical retinoid cycle in rods twilight vision                        | \n",
       "| 794 | Visual phototransduction                                                    | \n",
       "| 891 | Retinoid metabolism and transport                                           | \n",
       "| 933 | Metabolism of fat_soluble vitamins                                          | \n",
       "| ⋮ | ⋮ | \n",
       "| 32704 | Phase 1 _ Functionalization of compounds                                  | \n",
       "| 32796 | Xenobiotics                                                               | \n",
       "| 32816 | Aflatoxin activation and detoxification                                   | \n",
       "| 32837 | Glutathione conjugation                                                   | \n",
       "| 32875 | Glucuronidation                                                           | \n",
       "| 32895 | Cytosolic sulfonation of small molecules                                  | \n",
       "| 32913 | Synthesis and interconversion of nucleotide di_ and triphosphates         | \n",
       "| 32938 | Detoxification of Reactive Oxygen Species                                 | \n",
       "| 32971 | Metabolism of porphyrins                                                  | \n",
       "| 32988 | Purine metabolism                                                         | \n",
       "| 33022 | Metabolism of nucleotides                                                 | \n",
       "| 33109 | Pyrimidine metabolism                                                     | \n",
       "| 33133 | Complex I biogenesis                                                      | \n",
       "| 33182 | Respiratory electron transport                                            | \n",
       "| 33270 | Formation of ATP by chemiosmotic coupling                                 | \n",
       "| 33286 | Glyoxylate metabolism and glycine degradation                             | \n",
       "| 33311 | Pyruvate metabolism                                                       | \n",
       "| 33338 | Pyruvate metabolism and Citric Acid TCA cycle                             | \n",
       "| 33386 | Citric acid cycle TCA cycle                                               | \n",
       "| 33405 | Histidine lysine phenylalanine tyrosine proline and tryptophan catabolism | \n",
       "| 33446 | Mitochondrial tRNA aminoacylation                                         | \n",
       "| 33467 | tRNA Aminoacylation                                                       | \n",
       "| 33509 | Cytosolic tRNA aminoacylation                                             | \n",
       "| 33533 | tRNA modification in the nucleus and cytosol                              | \n",
       "| 33572 | Branched_chain amino acid catabolism                                      | \n",
       "| 33595 | Phase 4 _ resting membrane potential                                      | \n",
       "| 33614 | Mitochondrial translation initiation                                      | \n",
       "| 33698 | Mitochondrial translation termination                                     | \n",
       "| 33782 | Mitochondrial translation                                                 | \n",
       "| 33872 | Mitochondrial translation elongation                                      | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "      term                                                                       \n",
       "1     Fertilization                                                              \n",
       "27    Reproduction                                                               \n",
       "53    Miscellaneous transport and binding events                                 \n",
       "71    Uptake and actions of bacterial toxins                                     \n",
       "98    Dopamine Neurotransmitter Release Cycle                                    \n",
       "121   Serotonin Neurotransmitter Release Cycle                                   \n",
       "139   Norepinephrine Neurotransmitter Release Cycle                              \n",
       "157   Acetylcholine Neurotransmitter Release Cycle                               \n",
       "174   Glutamate Neurotransmitter Release Cycle                                   \n",
       "198   Neurotransmitter Release Cycle                                             \n",
       "249   GABA synthesis release reuptake and degradation                            \n",
       "268   Na+slashCl_ dependent neurotransmitter transporters                        \n",
       "287   Amine compound SLC transporters                                            \n",
       "317   Transport of inorganic cationsslashanions and amino acidsslasholigopeptides\n",
       "416   Amino acid transport across the plasma membrane                            \n",
       "447   Amino acid and oligopeptide SLC transporters                               \n",
       "496   Zinc transporters                                                          \n",
       "513   Metal ion SLC transporters                                                 \n",
       "538   Amino acid synthesis and interconversion transamination                    \n",
       "563   Transport of vitamins nucleosides and related molecules                    \n",
       "601   Disorders of transmembrane transporters                                    \n",
       "616   Surfactant metabolism                                                      \n",
       "645   ABC transporters in lipid homeostasis                                      \n",
       "662   ABC_family proteins mediated transport                                     \n",
       "707   Signaling by Retinoic Acid                                                 \n",
       "749   RA biosynthesis pathway                                                    \n",
       "771   The canonical retinoid cycle in rods twilight vision                       \n",
       "794   Visual phototransduction                                                   \n",
       "891   Retinoid metabolism and transport                                          \n",
       "933   Metabolism of fat_soluble vitamins                                         \n",
       "⋮     ⋮                                                                          \n",
       "32704 Phase 1 _ Functionalization of compounds                                   \n",
       "32796 Xenobiotics                                                                \n",
       "32816 Aflatoxin activation and detoxification                                    \n",
       "32837 Glutathione conjugation                                                    \n",
       "32875 Glucuronidation                                                            \n",
       "32895 Cytosolic sulfonation of small molecules                                   \n",
       "32913 Synthesis and interconversion of nucleotide di_ and triphosphates          \n",
       "32938 Detoxification of Reactive Oxygen Species                                  \n",
       "32971 Metabolism of porphyrins                                                   \n",
       "32988 Purine metabolism                                                          \n",
       "33022 Metabolism of nucleotides                                                  \n",
       "33109 Pyrimidine metabolism                                                      \n",
       "33133 Complex I biogenesis                                                       \n",
       "33182 Respiratory electron transport                                             \n",
       "33270 Formation of ATP by chemiosmotic coupling                                  \n",
       "33286 Glyoxylate metabolism and glycine degradation                              \n",
       "33311 Pyruvate metabolism                                                        \n",
       "33338 Pyruvate metabolism and Citric Acid TCA cycle                              \n",
       "33386 Citric acid cycle TCA cycle                                                \n",
       "33405 Histidine lysine phenylalanine tyrosine proline and tryptophan catabolism  \n",
       "33446 Mitochondrial tRNA aminoacylation                                          \n",
       "33467 tRNA Aminoacylation                                                        \n",
       "33509 Cytosolic tRNA aminoacylation                                              \n",
       "33533 tRNA modification in the nucleus and cytosol                               \n",
       "33572 Branched_chain amino acid catabolism                                       \n",
       "33595 Phase 4 _ resting membrane potential                                       \n",
       "33614 Mitochondrial translation initiation                                       \n",
       "33698 Mitochondrial translation termination                                      \n",
       "33782 Mitochondrial translation                                                  \n",
       "33872 Mitochondrial translation elongation                                       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pebba(file_in, gmt_file,force = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
